### PR TITLE
routing+migration32: update migration 32 to use pure TLV encoding for mission control results

### PR DESCRIPTION
**Pull request by ellemouton**
In this commit, we update an existing migration which at the time of
writing has not been included in a release. We update it so that it
converts the format used for MissionControl result encoding to use pure
TLV instead. The 3 structs that have been updated are: `mcHop`,
`mcRoute` and `paymentResult`.

### CODE REVIEW

File: channeldb/migration32/migration_test.go
Comment (guggero) [inline channeldb/migration32/migration_test.go:145]: nit: could define test hops as their own top-level variable to avoid some of the indentation. Would probably also be a bit easier to review.

File: channeldb/migration32/mission_control_store.go
Comment (guggero) [inline channeldb/migration32/mission_control_store.go:221]: nit: since we're passing a typed argument into this function, the second type parameter can be inferred by the compiler. So you can remote it to make things a bit more compact (here and in several other places).

File: channeldb/migration32/mission_control_store.go
Comment (guggero) [inline channeldb/migration32/mission_control_store.go:230]: nit: could use `fn.Map(extractMCHop, hops)`.

File: routing/missioncontrol.go
Comment (guggero) [inline routing/missioncontrol.go:266]: The reason we don't start at type 0 here is because of the ID which is part of the key and not the actual data, I assume? Could be worth a tiny comment.

File: routing/result_interpretation.go
Comment (guggero) [inline routing/result_interpretation.go:121]: nit: Took me a bit to grok this...
How about:
```
	var failMsg lnwire.FailureMessage
	failure.WhenSome(func(f failureMessage) {
		failMsg = f.FailureMessage
	})
```

**Review by guggero — APPROVED** [review]
tACK, LGTM :tada: 

Just a couple of nits and suggestions.

File: routing/result_interpretation.go
Comment (bitromortac) [inline routing/result_interpretation.go:111]: maybe one could use `ElimOption` to handle both cases in there, the downside would be that both branches are indented one level

**Review by bitromortac — COMMENTED** [review]
Looks great :zap:, will do some more tests.

File: channeldb/migration32/migration_test.go
Comment (ellemouton) [reply channeldb/migration32/migration_test.go:194]: good idea - updated 👍

File: routing/missioncontrol.go
Comment (ellemouton) [reply routing/missioncontrol.go]: ah - good point. This was an oversight cause I initially had ID as type 0. But then realised it is not actually serialised. Have updated so that timeFwd is type 0 and so on 👍

File: routing/missioncontrol_store.go
Comment (ellemouton) [reply routing/missioncontrol_store.go:571]: thanks :)

File: routing/result_interpretation.go
Comment (ellemouton) [reply routing/result_interpretation.go:111]: did you try do this and come up with a nice way? 
I gave it a try now but I dont find this very easy to read:

```
         idx := fn.ElimOption(errSourceIdx, func() int {
		i.processPaymentOutcomeUnknown(rt)

		return -1
	}, func(u uint8) int {
		return int(u)
	})
	if idx == -1 {
		return
	}
```

but maybe you had some other way in mind?

File: channeldb/migration32/mission_control_store.go
Comment (ellemouton) [reply channeldb/migration32/mission_control_store.go:218]: indeeeed - thanks!

File: channeldb/migration32/mission_control_store.go
Comment (ellemouton) [reply channeldb/migration32/mission_control_store.go]: ooo nice

**Review by ellemouton — COMMENTED** [review]
Thanks yall, updated

File: routing/result_interpretation.go
Comment (bitromortac) [reply routing/result_interpretation.go:111]: I thought about it using it this way (with a dummy return value), not sure how nice it is :)
```go
	fn.ElimOption(
		errSourceIdx,
		func() any {
			i.processPaymentOutcomeUnknown(rt)
			return nil
		},
		func(uidx uint8) any {
			idx := int(uidx)

			failMsg := failure.UnwrapOr(failureMessage{nil}).FailureMessage

			// If the payment was to a blinded route and we received
			// an error from after the introduction point, handle
			// this error separately - there has been a protocol
			// violation from the introduction node. This penalty
			// applies regardless of the error code that is
			// returned.
			introIdx, isBlinded := introductionPointIndex(rt)
			if isBlinded && introIdx < idx {
				i.processPaymentOutcomeBadIntro(
					rt, introIdx, idx,
				)

				return nil
			}

			switch idx {
			// We are the source of the failure.
			case 0:
				i.processPaymentOutcomeSelf(rt, failMsg)

			// A failure from the final hop was received.
			case len(rt.hops.Val):
				i.processPaymentOutcomeFinal(rt, failMsg)

			// An intermediate hop failed. Interpret the outcome,
			// update reputation and try again.
			default:
				i.processPaymentOutcomeIntermediate(
					rt, idx, failMsg,
				)
			}

			return nil
		},
	)
```

Comment (ellemouton) [reply routing/result_interpretation.go:111]: are you ok with me keeping it as is?

**Review by bitromortac — COMMENTED** [review]
Very close :rocket:, just a question if we want to also remove the success bool, which may be worth it because we do a migration.

I tested the migration as well and tested what happens if one runs with the old migration in master and then updates to the new migration, which will probably lead to an error like:

`[ERR] LTND lnd.go:159: Shutting down because error in main method: unable to create server: can't create mission control manager: tlv stream is not canonical`. This would require a manual deletion of the results bucket or the insertion of a temporary `purge` call at in `initMissionControl`.

**Comment by ellemouton** [comment]
Thanks @bitromortac  

> I tested the migration as well and tested what happens if one runs with the old migration in master and then updates to the new migration, which will probably lead to an error like:

Yes we dont support things if they were running on latest master. If they were, they can do a clear-mc call manually before starting again with this version. If we were supporting on latest master, then we would have added this as a new migration instead.

File: routing/result_interpretation.go
Comment (ellemouton) [reply routing/result_interpretation.go]: thanks - struggling to get my editor to do the thing....

File: routing/result_interpretation.go
Comment (ellemouton) [reply routing/result_interpretation.go:638]: cool yeah think im gonna leave this one just cause of the mismatch in signatures. So would need to put the full serialisation and deserialisation code here which makes it harder to read given the TLV specific wrapping that happens here

File: routing/result_interpretation.go
Comment (ellemouton) [reply routing/result_interpretation.go:83]: hmmm I dont think they are duplicates/redundancies. There are times when we want to report a failure but we don't know the index of the failure. See [this call](https://github.com/lightningnetwork/lnd/blob/acbb33bb7b11e2dd3f2f85fe14fd3ff087212fc0/routing/payment_lifecycle.go#L889) where we are reporting a failure but we set the source index and failure messages to nil

**Review by ellemouton — COMMENTED** [review]
thanks @bitromortac  - updated!

File: routing/result_interpretation.go
Comment (bitromortac) [reply routing/result_interpretation.go:83]: right, but in the end we only have successes or failures, so we could combine the failure index and failure message into an optional TLV sub-struct whose presence encodes if it was a failure? but non-blocking, as it's not a huge gain in savings, but it would enforce that we can't have a success and failures set in the same result

**Review by bitromortac — APPROVED** [review]
> Yes we dont support things if they were running on latest master. If they were, they can do a clear-mc call manually before starting again with this version. If we were supporting on latest master, then we would have added this as a new migration instead.

Right, just in case anybody updates without knowing about the migrations beforehand, it's possible to get stuck, because one can't `resetmc` then due to the rpc server not being available, just wanted to have an answer for that case.

LGTM :zap: (pending some clarification on a suggestion)

File: routing/result_interpretation.go
Comment (ellemouton) [reply routing/result_interpretation.go:83]: ah! good point - cool will update 👍

**Comment by ellemouton** [comment]
> Right, just in case anybody updates without knowing about the migrations beforehand, it's possible to get stuck, because one can't resetmc then due to the rpc server not being available, just wanted to have an answer for that case.

Yeah, I think in that case they need to manually drop the bucket (kvdb) or write a cascading delete (sql). Let's defs confirm with @Roasbeef  though. 

I just think that if we do want to support users in this state, then we should have gone with adding a separate migration all together. If we made the decision about editing an existing migration, then to me that means not supporting this state.

**Comment by ellemouton** [comment]
@bitromortac  - in the mean time, I've updated the PR to use the implicit success/fail struct you suggested - mind having a look at the new structure? Once I get the ACK from you, then I'll ask Oli to also take another look given the structure is quite different now

**Review by bitromortac — COMMENTED** [review]
The latest changes look good :pray:, leaving some comments, but will also do another round of testing.

File: routing/result_interpretation.go
Comment (ellemouton) [reply routing/result_interpretation.go:90]: agreed but such a version is not available right now right?

Comment (ellemouton) [reply routing/result_interpretation.go:90]: oh cool!

File: routing/missioncontrol.go
Comment (ellemouton) [reply routing/missioncontrol.go:275]: yeah not that i can think of 🤔

File: routing/result_interpretation.go
Comment (ellemouton) [reply routing/result_interpretation.go]: thanks - updated

File: routing/missioncontrol.go
Comment (ellemouton) [reply routing/missioncontrol.go:275]: because we want the presence of `paymentFailure` to mean failure but then have to also cater for when we dont know details of the failure

File: routing/missioncontrol_store_test.go
Comment (guggero) [inline routing/missioncontrol_store_test.go:93]: Can we also add a test with a failure that has no details? Just to show that works too (didn't see one, but perhaps I didn't look hard enough).

**Review by guggero — APPROVED** [review]
re-tACK, LGTM :tada:

File: routing/missioncontrol_store_test.go
Comment (ellemouton) [reply routing/missioncontrol_store_test.go:93]: indeed - good idea! added

**Pull request merged**
